source/common/http/conn_pool_grid.cc

```diff
@@ -0,0 +1,185 @@
+                                                     ConnectionPool::Callbacks& callbacks)
+    : grid_(grid), decoder_(decoder), inner_callbacks_(callbacks), pool_it_(pool_it) {
+  cancellable_ = pool().newStream(decoder_, *this);
+}
+
+void ConnectivityGrid::WrapperCallbacks::onPoolFailure(
+    ConnectionPool::PoolFailureReason reason, absl::string_view transport_failure_reason,
+    Upstream::HostDescriptionConstSharedPtr host) {
+  // When a connection fails, see if there is a lower priority pool to attempt
+  // connecting with.
+  absl::optional<PoolIterator> next_pool = grid_.nextPool(pool_it_);
+  if (next_pool.has_value()) {
+    pool_it_ = next_pool.value();
+    cancellable_ = pool().newStream(decoder_, *this);
+    return;
+  }
+
+  // If this point is reached, all pools have been tried. Pass the pool failure up to the
+  // original caller.
+  ConnectionPool::Callbacks& callbacks = inner_callbacks_;
+  delete this;
+  callbacks.onPoolFailure(reason, transport_failure_reason, host);
+}
+
+void ConnectivityGrid::WrapperCallbacks::onPoolReady(RequestEncoder& encoder,
+                                                     Upstream::HostDescriptionConstSharedPtr host,
+                                                     const StreamInfo::StreamInfo& info,
+                                                     absl::optional<Http::Protocol> protocol) {
+  // Right now, connections are tried serially, so any successful stream
+  // creation should be passed up to the original caller.
+  ConnectionPool::Callbacks& callbacks = inner_callbacks_;
+  delete this;
+  return callbacks.onPoolReady(encoder, host, info, protocol);
+}
+
+void ConnectivityGrid::WrapperCallbacks::cancel(Envoy::ConnectionPool::CancelPolicy cancel_policy) {
+  // If the newStream caller cancels the stream request, pass the cancellation on
+  // to the active pool.
+  cancellable_->cancel(cancel_policy);
+  delete this;
+}
+
+ConnectivityGrid::ConnectivityGrid(
+    Event::Dispatcher& dispatcher, Random::RandomGenerator& random_generator,
+    Upstream::HostConstSharedPtr host, Upstream::ResourcePriority priority,
+    const Network::ConnectionSocket::OptionsSharedPtr& options,
+    const Network::TransportSocketOptionsSharedPtr& transport_socket_options,
+    Upstream::ClusterConnectivityState& state, TimeSource& time_source,
+    ConnectivityOptions connectivity_options)
+    : dispatcher_(dispatcher), random_generator_(random_generator), host_(host),
+      priority_(priority), options_(options), transport_socket_options_(transport_socket_options),
+      state_(state), time_source_(time_source) {
+  // TODO(#15649) support v6/v4, WiFi/cellular.
+  ASSERT(connectivity_options.protocols_.size() == 3);
+  ASSERT(contains(connectivity_options.protocols_,
+                  {Http::Protocol::Http11, Http::Protocol::Http2, Http::Protocol::Http3}));
+}
+
+ConnectivityGrid::~ConnectivityGrid() {
+  // Ignore drained callbacks during teardown.
+  destroying_ = true;
```

Comment:
Can you elaborate on this? Unless I'm misunderstanding something, accessing the value of `destroying_` after the destructor returns is undefined behavior

Reply:
Ah thanks for that and the comment